driver-core/src/main/com/mongodb/client/model/densify/DensifyRange.java

```diff
@@ -0,0 +1,164 @@
+        return new DensifyConstructibleBson(new Document("bounds", "partition")
+                .append("step", notNull("step", step)));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents a single interval [l, u).
+     *
+     * @param l The lower bound.
+     * @param u The upper bound.
+     * @param step The step.
+     * @return The requested {@link DensifyRange}.
+     */
+    static NumberDensifyRange rangeWithStep(final Number l, final Number u, final Number step) {
+        notNull("l", l);
+        notNull("u", u);
+        notNull("step", step);
+        return new DensifyConstructibleBson(new Document("bounds", asList(l, u))
+                .append("step", notNull("step", step)));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents an interval with the smallest BSON {@link BsonType#DATE_TIME Date} value
+     * of the {@linkplain Aggregates#densify(String, DensifyRange, DensifyOptions) field}
+     * in the sequence of documents being its lower bound, and the largest value being the upper bound.
+     *
+     * @param step The step.
+     * @param unit The unit in which the {@code step} is specified.
+     * @return The requested {@link DensifyRange}.
+     */
+    static DateDensifyRange fullRangeWithStep(final long step, final MongoTimeUnit unit) {
+        notNull("unit", unit);
+        return new DensifyConstructibleBson(new BsonDocument("bounds", new BsonString("full"))
+                .append("step", new BsonInt64(step))
+                .append("unit", new BsonString(unit.value())));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents an interval with the smallest BSON {@link BsonType#DATE_TIME Date} value
+     * of the {@linkplain Aggregates#densify(String, DensifyRange, DensifyOptions) field}
+     * in the {@linkplain DensifyOptions#partitionByFields(Iterable) partition} of documents being its lower bound,
+     * and the largest value being the upper bound.
+     *
+     * @param step The step.
+     * @param unit The unit in which the {@code step} is specified.
+     * @return The requested {@link DensifyRange}.
+     */
+    static DateDensifyRange partitionRangeWithStep(final long step, final MongoTimeUnit unit) {
+        notNull("unit", unit);
+        return new DensifyConstructibleBson(new BsonDocument("bounds", new BsonString("partition"))
+                .append("step", new BsonInt64(step))
+                .append("unit", new BsonString(unit.value())));
+    }
+
+    /**
+     * Returns a {@link DensifyRange} that represents a single interval [l, u).
+     *
+     * @param l The lower bound.
+     * @param u The upper bound.
+     * @param step The step.
+     * @param unit The unit in which the {@code step} is specified.
+     * @return The requested {@link DensifyRange}.
+     */
+    static DateDensifyRange rangeWithStep(final Instant l, final Instant u, final long step, final MongoTimeUnit unit) {
```

Comment:
Returning a more specific type has the following benefits:

- We may introduce type-specific methods ("setters"), thus expressing that an optional field is present on `DateDensifyRange` and not present on `NumberDensifyRange`.
- We may accept a more specific type as a parameter in some methods, thus guiding a user towards using the API correctly, instead of expressing the type restrictions in the docs. For example, we can accept `DateDensifyRange`, instead of accepting `DensifyRange` and additionally documenting that only objects created as date ranges are allowed to be used.

Both of the above benefits were used in the `$search`/`$searchMeta` API. In this PR I am just following the idea in case it becomes handy in the future (we won't be able to return a more specific type in the future without breaking binary compatibility).